### ISSUE TITLE
Extensibility problems allowing dsls to break each other

### ISSUE THREAD

**Issue by ernesto-jimenez** [goadesign/goa#268]
## Problem

From the authoring documentation:

> The DSL must register its root objects in the DSL engine package Roots variable for the engine to execute the DSL.

However, there are several issues with the `dslengine.Roots` variable and the current DSL and generators which harm extensibility:

### It's currently initialised by [`apidsl`'s init](https://github.com/goadesign/goa/blob/master/design/apidsl/init.go#L96)

If a different DSL is initialised before `apidsl`, it'll be removed from `dslengine.Roots` on `apidsl`'s init.

```go
% grep -R 'dslengine.Roots =' .
./design/apidsl/attribute_test.go:              dslengine.Roots = append(dslengine.Roots, t)
./design/apidsl/init.go:        dslengine.Roots = []dslengine.Root{api}
./design/apidsl/media_type.go:          dslengine.Roots = append(dslengine.Roots, design.GeneratedMediaTypes)
./design/types.go:              dslengine.Roots = append(dslengine.Roots, GeneratedMediaTypes)
```

### Current generators would panic depending on the order of `dslengine.Roots` 

Generators currently expect `design.APIDefinition` to be the first argument in `Roots`

```
% grep -R 'roots' goagen -n
goagen/gen_app/generator.go:22:func Generate(roots []interface{}) (files []string, err error) {
goagen/gen_app/generator.go:23: api := roots[0].(*design.APIDefinition)
goagen/gen_client/generator.go:25:func Generate(roots []interface{}) (files []string, err error) {
goagen/gen_client/generator.go:26:      api := roots[0].(*design.APIDefinition)
goagen/gen_gen/command_test.go:80:func Generate(roots []interface{}) ([]string, error) {
goagen/gen_js/generator.go:25:func Generate(roots []interface{}) (files []string, err error) {
goagen/gen_js/generator.go:26:  api := roots[0].(*design.APIDefinition)
goagen/gen_main/generator.go:25:func Generate(roots []interface{}) (files []string, err error) {
goagen/gen_main/generator.go:26:        api := roots[0].(*design.APIDefinition)
goagen/gen_schema/generator.go:21:func Generate(roots []interface{}) (files []string, err error) {
goagen/gen_schema/generator.go:22:      api := roots[0].(*design.APIDefinition)
goagen/gen_swagger/generator.go:20:func Generate(roots []interface{}) (files []string, err error) {
goagen/gen_swagger/generator.go:21:     api := roots[0].(*design.APIDefinition)
```

## Requirements

The aim of the solution is to avoid DSLs to override each other *and* prevent generators from making assumptions on the list of definitions to increase compatibility between DSLs.

### Avoiding DSLs overriding each other

This one is simple, the array of `dslengine` `Root`s should be unexported and initialised by `dslengine`. There would be an interface to 

### Prevent generators from making assumptions on the list of definitions

`dslengine` should have an API to root definitions of an specific type rather than taking a raw `roots []interface{}` that would enable developers to make assumptions on the structure of the array.

## Proposed solution

Something like this:

```go
package dslengine

// [...]

// Roots return the root definitions from the DSLs in use
func Roots() RootDefinitions {
}

type RootDefinitions interface {
	// Register adds a new definition
	Register(Root)
	// Walk goes through the list registered definitions and calls the given function.
	// If the function returns an error, the walk will be stopped and the error will be returned.
	Walk(func(Root) error) error
}
```

Current generators would then work like this:

```go
func Generate(defs dslengine.RootDefinitions) ([]string, error) {
	var api *design.APIDefinition
	err := defs.Walk(func(d interface{}) error {
		if a, ok := d.(*design.APIDefinition); ok && api == nil {
			api = a
		} else if ok {
			return errors.New("only one APIDefinition can be registered")
		}
	})
	if err != nil {
		return nil, err
	}
	if api == nil {
		return nil, errors.New("no APIDefinition registered in the API design")
	}
}
```

That could obviously be refactored into a `design.FindAPIDefinition(dslengine.RootDefinitions) (*design.APIDefinition, error)` simplifying the `Generate` functions of the current generators.

**Comment by raphael** [comment]
Looks great, thanks!

**Comment by bketelsen** [comment]
@ernesto-jimenez are you still working on this fix?

**Comment by raphael** [comment]
A few notes that surfaced while taking a closer look:

* Using a function that returns a set of root definitions for use in tests does not work because the DSL operates on package variables.
* Instead of having this embedded array in a struct for RootDefinitions it seems simpler to have it be a slice and have the `Register` and `Walk` methods be package functions (given the above).
* There should be an `Reset` package function that tests can use to reset the content of the private `roots` package variable.

**Comment by ernesto-jimenez** [comment]
> * Using a function that returns a set of root definitions for use in tests does not work because the DSL operates on package variables.

You can have functions to register new root definitions or get the latest. I believe that's how I implemented it in #337 instead of package variables.

> * Instead of having this embedded array in a struct for RootDefinitions it seems simpler to have it be a slice and have the Register and Walk methods be package functions (given the above).

The idea is to avoid exposing an array/slice to the DSLs, since they can then interfere with each other.

> * There should be an Reset package function that tests can use to reset the content of the private roots package variable.

If there's a `Reset`, it should be behind a `test` build tag so it's only available for testing, to make sure DSLs do not call it.

No DSL package should be able to call `Reset` during `init`. It would be assuming it's the first package to be imported and there might be another package imported and initialised before whose registrations are overridden by that package calling `Reset`.

> * FindAPIDefinition is not strictly needed as the apidsl package keeps the root in the Design package variable on top of registering it with the dslengine package.

Is this just the current behaviour or will it always be guaranteed no matter which plugins are being used?

**Comment by raphael** [comment]
@ernesto-jimenez  sorry I had missed your comments - just stumbled on them... 

* The pattern of using a function that returns a set of root definitions for use in tests does not work because the tests run the DSL which in turns operate on the package variable - tests have no control on that. 

* The roots are not exposed - instead there is a `Register` method like you introduced.

* I like the idea of hiding the `Reset` behind a `test` build tag, thanks!

* Each DSL package is free to decide how to expose (or not) its root definitions.
